README
NAME
    MooseX::MarkAsMethods - Mark overload code symbols as methods

SYNOPSIS
        package Foo;
        use Moose;

        # mark overloads as methods and wipe other non-methods
        use MooseX::MarkAsMethods autoclean => 1;

        # define overloads, etc as normal

        package Baz;
        use Moose::Role
        use MooseX::MarkAsMethods autoclean => 1;

        # overloads defined in a role will "just work" when the role is
        # composed into a class

        # additional methods generated outside Class::MOP/Moose can be marked, too
        __PACKAGE__->meta->mark_as_method('foo');

        package Bar;
        use Moose;

        # order is important!
        use namespace::autoclean;
        use MooseX::MarkAsMethods;

        # ...

DESCRIPTION
    MooseX::MarkAsMethods allows one to easily mark certain functions as
    Moose methods. This will allow other packages such as
    namespace::autoclean to operate without blowing away your overloads.
    After using MooseX::MarkAsMethods your overloads will be recognized by
    Class::MOP as being methods, and class extension as well as composition
    from roles with overloads will "just work".

    By default we check for overloads, and mark those functions as methods.

    If 'autoclean => 1' is passed to import on use'ing this module, we will
    invoke namespace::autoclean to clear out non-methods.

TRAITS APPLIED
    use'ing this package causes a trait to be applied to your metaclass (for
    both roles and classes), that provides a mark_as_method() method. You
    can use this to mark newly generated methods at runtime (e.g. during
    class composition) that some other package has created for you.

    mark_as_method() is invoked with one or more names to mark as a method.
    We die on any error (e.g. name not in symbol table, already a method,
    etc). e.g.

        __PACKAGE__->meta->mark_as_method('newly_generated');

    e.g. say you have some sugar from another package that creates accessors
    of some sort; you could mark them as methods via a method modifier:

        # called as __PACKAGE__->foo_generator('name', ...)
        after 'foo_generator' => sub {

            shift->meta->mark_as_method(shift);
        };

IMPLICATIONS FOR ROLES
    Using MooseX::MarkAsMethods in a role will cause Moose to track and
    treat your overloads like any other method defined in the role, and
    things will "just work". That's it.

CAVEATS
  meta->mark_as_method()
    You almost certainly don't need or want to do this. CMOP/Moose are
    fairly good about determining what is and what isn't a method, but not
    perfect. Before using this method, you should pause and think about why
    you need to.

  namespace::autoclean
    As currently implemented, we run our "method maker" at the end of the
    calling package's compile scope (B::Hooks::EndOfScope). As
    namespace::autoclean does the same thing, it's important that if
    namespace::autoclean is used that it be use'd BEFORE
    MooseX::MarkAsMethods, so that its end_of_scope block is run after ours.

    e.g.

        # yes!
        use namespace::autoclean;
        use MooseX::MarkAsMethods;

        # no -- overloads will be removed
        use namespace::autoclean;
        use MooseX::MarkAsMethods;

    The easiest way to invoke this module and clean out non-methods without
    having to worry about ordering is:

        use MooseX::MarkAsMethods autoclean => 1;

SEE ALSO
    overload, B::Hooks::EndOfScope, namespace::autoclean, Class::MOP, Moose.

    MooseX::Role::WithOverloading does allow for overload application from
    roles, but it does this by copying the overload symbols from the (not
    namespace::autoclean'ed role) the symbols handing overloads during class
    composition; we work by marking the overloads as methods and letting
    CMOP/Moose handle them.

AUTHOR
    Chris Weyl, "<cweyl at alumni.drew.edu>"

BUGS
    Please report any bugs or feature requests to "bug-moosex-markasmethods
    at rt.cpan.org", or through the web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=MooseX-MarkAsMethods>.

TODO
    Additional testing is required, particularly where namespace::autoclean
    is also being used.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc MooseX::MarkAsMethods

    You can also look for information at:

    *   RT: CPAN's request tracker

        <http://rt.cpan.org/NoAuth/Bugs.html?Dist=MooseX-MarkAsMethods>

    *   AnnoCPAN: Annotated CPAN documentation

        <http://annocpan.org/dist/MooseX-MarkAsMethods>

    *   CPAN Ratings

        <http://cpanratings.perl.org/d/MooseX-MarkAsMethods>

    *   Search CPAN

        <http://search.cpan.org/dist/MooseX-MarkAsMethods/>

COPYRIGHT & LICENSE
    Copyright (c) 2009, 2010, Chris Weyl "<cweyl@alumni.drew.edu>".

    This library is free software; you can redistribute it and/or modify it
    under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation; either version 2.1 of the License, or (at
    your option) any later version.

    This library is distributed in the hope that it will be useful, but
    WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS OR A PARTICULAR PURPOSE.

    See the GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with this library; if not, write to the

        Free Software Foundation, Inc.,
        59 Temple Place, Suite 330,
        Boston, MA  02111-1307 USA



Changes
Changes for MooseX-MarkAsMethods

{{$NEXT}}

0.15      2012-05-30 17:34:11 America/Los_Angeles
	* Document that overload declarations must use the anon sub style for
	  methods to resolve correctly.
	* No code/functionality changes.

0.14      2011-11-03 21:08:52 America/Los_Angeles
	* honor the "into => 'X'" flag Sub::Exporter recognizes

0.13      2011-09-17 10:08:31 America/Los_Angeles
	* drop Perl6::Junction dependency; we don't use it (anymore?)

0.12      2011-09-16 11:10:21 America/Los_Angeles
	* restore Moose minimum version as 0.94

0.11      2011-03-31 15:07:38 America/Los_Angeles
    * Correct POD error
    * Switch to Dist::Zilla

0.10 - Mon Jun 14 2010
	* Add author_requires for modules needed by xt/ bits.
	* MooseX/MarkAsMethods.pm was being a touch misleading, when asking for
	  0.90 of Moose.  This fell out of sync with the requirements in
	  Makefile.PL, and as such has been removed.

0.09 - Fri Jun 11 2010
	* push Moose required version (rt#58309)

0.08 - Tue May 25 2010
	* Remove "inc" from MANIFEST.SKIP...  Not entirely sure how that snuck in
	  there.

0.07 - Sat May 22 2010
	* Use wrap_method_body() rather than using our own method metaclass
	  subclass, and drop said subclass
	* Apply a trait to the class/role metaclass that provides a
	  mark_as_method() method, and use that do do our method marking
	* Initial test-cases for after-the-fact method marking
	* Documentation updates w.r.t. the above

0.06 - Tue May 11 2010
	* Added a couple additional test cases
	* Cleaned up our dependencies (*sigh*)

0.05 - Mon Dec 14 20:15:32 2009 -0800
  Chris Weyl <cweyl@alumni.drew.edu> (2):
    make Perl6::Junction required for testing
    Updating $VERSION to 0.05

0.04 - Fri Dec 4 17:04:23 2009 -0800
  Chris Weyl <cweyl@alumni.drew.edu> (4):
    use namespace::clean->import()
    abstract out o/l testing into check_overloads()
    unmark author boilerplate tests as TODO
    Updating $VERSION to 0.04

0.03 - Wed Nov 25 22:15:47 2009 -0800
  Chris Weyl <cweyl@alumni.drew.edu> (2):
    .pm updates (autoclean)
    Rework and enlarge test suite.



t/funcs.pm
my @sugar = qw{ has around augment inner before after blessed confess };

sub check_sugar_removed_ok {
    my $t = shift @_;

    # check some (not all) Moose sugar to make sure it has been cleared
    #my @sugar = qw{ has around augment inner before after blessed confess };
    ok !$t->can($_) => "$t cannot $_" for @sugar;

    return;
}

sub check_sugar_ok {
    my $t = shift @_;

    # check some (not all) Moose sugar to make sure it has been cleared
    #my @sugar = qw{ has around augment inner before after blessed confess };
    ok $t->can($_) => "$t can $_" for @sugar;

    return;
}

sub make_and_check {
    #my $class = shift @_;
    my ($class, $roles, $atts) = @_;

    my $t = $class->new;
    isa_ok  $t, $class;

    # do our class checks: meta, roles, attributes
    meta_ok $class;
    does_ok $class => $_ for @$roles;
    has_attribute_ok $class => $_ for @$atts;

    return $t;
}

sub check_overloads {
    my ($t, %overloads) = @_;

    die "We expect an instance of $t, not a classname"
        unless ref $t;

    my $class = ref $t;
    ok overload::Overloaded($class), "$class is subject to some overloads";

    for my $op (keys %overloads) {

        # check that Moose knows about it, overload knows about it, and that
        # it works the way we expect it to

        if ($t->meta->has_method("($op")) {

            # we have the method and are its originator
            pass "$class still has o/l method ($op";
        }
        else {

            # we have the method via inheriting, etc
            ok $t->meta->find_method_by_name("($op"), "$class inherits o/l method ($op";
        }

        ok overload::Method($t, $op), "overload claims $class has $op overloaded";
        is "$t", $overloads{$op}, "$class o/l returned the expected value";
    }

    return;
}

sub check_methods    { _check_methods(\&pass, \&fail, @_) }
sub check_no_methods { _check_methods(\&fail, \&pass, @_) }

sub _check_methods {
    my ($has, $not_has, $t, @methods) = @_;
    my $class = ref $t;

    for my $method (@methods) {

        # see if we have it directly...
        do { $has->("$class has method $method"); next }
            if $t->meta->has_method($method);

        # ... or via inheritance
        do { $has->("$class inherits method $method"); next }
            if $t->meta->find_method_by_name($method);

        # if we're here, it's a fail
        $not_has->("$class neither has nor inherits method $method");
    }

    return;
}

1;


t/lib/Simple.pm
#############################################################################
#
# 
#
# Author:  Chris Weyl (cpan:RSRCHBOY), <cweyl@alumni.drew.edu>
# Company: No company, personal work
# Created: 11/22/2009
#
# Copyright (c) 2009  <cweyl@alumni.drew.edu>
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
#############################################################################

package Simple;

use Moose;
use MooseX::MarkAsMethods;
use namespace::autoclean;

use overload q{""} => 'stringify', fallback => 1;

has id => (is => 'rw', isa => 'Str', default => 'bob');

sub stringify { shift->id }

__PACKAGE__->meta->make_immutable;
1;


t/lib/SimpleChild.pm
#############################################################################
#
# Test class. 
#
# Author:  Chris Weyl (cpan:RSRCHBOY), <cweyl@alumni.drew.edu>
# Company: No company, personal work
# Created: 11/25/2009
#
# Copyright (c) 2009  <cweyl@alumni.drew.edu>
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
#############################################################################

package SimpleChild;

use Moose;
use MooseX::MarkAsMethods;
extends 'Simple';

#use namespace::clean -except => 'meta';

our $VERSION = '0.001';

#sub stringify { 'bob2' }

__PACKAGE__->meta->make_immutable;

__END__


lib/MooseX/MarkAsMethods.pm
package MooseX::MarkAsMethods;

# ABSTRACT: Mark overload code symbols as methods

use warnings;
use strict;

use namespace::autoclean 0.12;

use B::Hooks::EndOfScope;
use Moose 0.94 ();
use Moose::Util::MetaRole;
use Moose::Exporter;

# debugging
#use Smart::Comments '###', '####';

{
    package MooseX::MarkAsMethods::MetaRole::MethodMarker;
    use Moose::Role;
    use namespace::autoclean;

    sub mark_as_method {
        my $self = shift @_;

        $self->_mark_as_method($_) for @_;
        return;
    }

    sub _mark_as_method {
        my ($self, $method_name) = @_;

        do { warn "$method_name is already a method!"; return }
            if $self->has_method($method_name);

        my $code = $self->get_package_symbol({
            name  => $method_name,
            sigil => '&',
            type  => 'CODE',
        });

        do { warn "$method_name not found as a CODE symbol!"; return }
            unless defined $code;

        $self->add_method($method_name =>
            $self->wrap_method_body(
                associated_metaclass => $self,
                name => $method_name,
                body => $code,
            ),
        );

        return;
    }
}

my ($import) = Moose::Exporter->build_import_methods(
    install => [ qw{ init_meta unimport } ],
    class_metaroles => {
        class => ['MooseX::MarkAsMethods::MetaRole::MethodMarker'],
    },
    role_metaroles => {
        role => ['MooseX::MarkAsMethods::MetaRole::MethodMarker'],
    },
);

sub import {
    #my ($class, @args) = @_;
    my $class = shift @_;

    # if someone is passing in Sub::Exporter-style initial hash, grab it
    my $exporter_opts;
    $exporter_opts = shift @_ if ref $_[0] && ref $_[0] eq 'HASH';
    my %args = @_;

    #my $target = $exporter_opts->{into} if $exporter_opts;
    #$target  ||= scalar caller;
    my $target
        = defined $exporter_opts && defined $exporter_opts->{into}
        ? $exporter_opts->{into}
        : scalar caller
        ;

    return if $target eq 'main';
    #$class->init_meta(for_class => $target);

    my $do_autoclean = delete $args{autoclean};

    on_scope_end {

        ### $target
        my $meta = Class::MOP::Class->initialize($target);

        ### metaclass: ref $meta
        my %methods   = map { ($_ => 1) } $meta->get_method_list;
        my %symbols   = %{ $meta->get_all_package_symbols('CODE') };
        my @overloads = grep { /^\(/ } keys %symbols;

        ### %methods
        ### %symbols
        ### @overloads

        foreach my $overload_name (@overloads) {

            next if $methods{$overload_name};

            ### marking as method: $overload_name
            $meta->mark_as_method($overload_name);
            $methods{$overload_name} = 1;
            delete $symbols{$overload_name};
        }

        return;
    };

    ### $do_autoclean
    namespace::autoclean->import(-cleanee => $target)
        if $do_autoclean;

    @_ = $exporter_opts ? ($exporter_opts, %args) : (%args);
    unshift @_, $class;

    ### @_
    goto &$import;
    return;
}

1;

__END__

=for Pod::Coverage init_meta

=head1 SYNOPSIS

    package Foo;
    use Moose;

    # mark overloads as methods and wipe other non-methods
    use MooseX::MarkAsMethods autoclean => 1;

    # define overloads, etc as normal
    use overload '""' => sub { shift->stringify };

    package Baz;
    use Moose::Role;
    use MooseX::MarkAsMethods autoclean => 1;

    # overloads defined in a role will "just work" when the role is
    # composed into a class; they MUST use the anon-sub style invocation
    use overload '""' => sub { shift->stringify };

    # additional methods generated outside Class::MOP/Moose can be marked, too
    use constant foo => 'bar';
    __PACKAGE__->meta->mark_as_method('foo');

    package Bar;
    use Moose;

    # order is important!
    use namespace::autoclean;
    use MooseX::MarkAsMethods;

    # ...

=head1 DESCRIPTION

MooseX::MarkAsMethods allows one to easily mark certain functions as Moose
methods.  This will allow other packages such as L<namespace::autoclean> to
operate without blowing away your overloads.  After using
MooseX::MarkAsMethods your overloads will be recognized by L<Class::MOP> as
being methods, and class extension as well as composition from roles with
overloads will "just work".

By default we check for overloads, and mark those functions as methods.

If C<autoclean =&gt; 1> is passed to import on using this module, we will invoke
namespace::autoclean to clear out non-methods.

=head1 TRAITS APPLIED

Using this package causes a trait to be applied to your metaclass (for both
roles and classes), that provides a mark_as_method() method.  You can use this
to mark newly generated methods at runtime (e.g. during class composition)
that some other package has created for you.

mark_as_method() is invoked with one or more names to mark as a method.  We die
on any error (e.g. name not in symbol table, already a method, etc).  e.g.

    __PACKAGE__->meta->mark_as_method('newly_generated');

e.g. say you have some sugar from another package that creates accessors of
some sort; you could mark them as methods via a method modifier:

    # called as __PACKAGE__->foo_generator('name', ...)
    after 'foo_generator' => sub {

        shift->meta->mark_as_method(shift);
    };

=head1 IMPLICATIONS FOR ROLES

Using MooseX::MarkAsMethods in a role will cause Moose to track and treat your
overloads like any other method defined in the role, and things will "just
work".  That's it.

Except...  note that due to the way overloads, roles, and Moose work, you'll
need to use the coderef or anonymous subroutine approach to overload
declaration, or things will not work as you expect.  Remember, we're talking
about _methods_ here, so we need to make it easy for L<overload> to find
the right method.  The easiest (and supported) way to do this is to create an
anonymous sub to wrap the overload method.

That is, this will work:

    # note method resolution, things will "just work"
    use overload '""' => sub { shift->stringify };

...and this will not:

    use overload '""' => 'stringify';

...and will result in an error message like:

    # wah-wah
    Can't resolve method "???" overloading """" in package "overload"

=head1 CAVEATS

=head2 Roles

See the "IMPLICATIONS FOR ROLES" section, above.

=head2 meta->mark_as_method()

B<You almost certainly don't need or want to do this.>  CMOP/Moose are fairly
good about determining what is and what isn't a method, but not perfect.
Before using this method, you should pause and think about why you need to.

=head2 namespace::autoclean

As currently implemented, we run our "method maker" at the end of the calling
package's compile scope (L<B::Hooks::EndOfScope>).  As L<namespace::autoclean>
does the same thing, it's important that if namespace::autoclean is used that
it be used BEFORE MooseX::MarkAsMethods, so that its end_of_scope block is
run after ours.

e.g.

    # yes!
    use namespace::autoclean;
    use MooseX::MarkAsMethods;

    # no -- overloads will be removed
    use MooseX::MarkAsMethods;
    use namespace::autoclean;

The easiest way to invoke this module and clean out non-methods without having
to worry about ordering is:

    use MooseX::MarkAsMethods autoclean => 1;

=head1 SEE ALSO

L<overload>, L<B::Hooks::EndOfScope>, L<namespace::autoclean>, L<Class::MOP>,
L<Moose>.

L<MooseX::Role::WithOverloading> does allow for overload application from
roles, but it does this by copying the overload symbols from the (not
L<namespace::autoclean>'ed role) the symbols handing overloads during class
composition; we work by marking the overloads as methods and letting
CMOP/Moose handle them.

=cut

